File: Python/run_l96_var.py
import numpy as np
import l96
from config_l96 import ns, F, dt, r, nt, \
        b, nw, ni, a, g_break, xt_fname, yo_fname, xf_fname

nc = nt // nw
def calc_cost(dx, b, dy, r):
    cost = 0.5 * np.dot(dx, dx) / b
    for j in range(dy.shape[1]):
        cost += 0.5 * np.dot(dy[:, j], dy[:, j]) / r
    return cost

f_xt = open(xt_fname, "rb")
f_yo = open(yo_fname, "rb")
x0 = np.fromfile(xf_fname, count=ns)

yo = np.zeros((ns, nw))
xb = np.zeros((ns, nw))
d = np.zeros((ns, nw))
l2 = np.zeros(nc)
ad = np.zeros(ns)

for k in range(nc):
    print(f"cycle={k+1}")
    xt0 = np.fromfile(f_xt, count=ns)
    for j in range(1, nw):
        xt = np.fromfile(f_xt, count=ns)
    for j in range(0, nw):
        yo[:, j] = np.fromfile(f_yo, count=ns)
    xb[:, 0] = x0
    for i in range(ni):
        for j in range(1, nw):
            xb[:, j] = l96.fom(xb[:, j-1], 1, dt, F)
        d = xb - yo
        ad = 0
        for j in range(nw-1, -1, -1):
            ad = l96.adm(xb[:, j], ad, 1, dt, F) + d[:, j] / r
        xb[:, 0] -= a * ad
        cost = calc_cost(xb[:, 0] - x0, b, d, r)
        if i == 0:
            cost_old = cost
        l2[k] = np.sqrt(np.mean((xb[:, 0] - xt0)**2))
        gnorm = np.dot(ad, ad)
        print(f"{i+1} J={cost:.2e} Jold={cost_old:.2e} g={gnorm:.2e} l2={l2[k]:.2e}")
        if gnorm < g_break:
            break
        if cost > cost_old:
            xb[:, 0] += a * ad
            l2[k] = np.sqrt(np.mean((xb[:, 0] - xt0)**2))
            break
        cost_old = cost
    x0 = l96.fom(xb[:, 0], nw + 1, dt, F)
f_xt.close()
f_yo.close()

l2.tofile("l96_var.dat")

File: R/run_l96_var.R
source("l96.R")
source("step.R")
source("config_l96.R")

nc <- nt %/% nw
calc.cost <- function(dx, b, dy, r) {
  cost <- 0.5 * t(dx) %*% dx / b
  for (j in 1:ncol(dy)) {
    cost <- cost + 0.5 * t(dy[, j]) %*% dy[, j] / r
  }
  as.numeric(cost)
}

con.true <- file(xt_fname, "rb")
con.obs <- file(yo_fname, "rb")
con <- file(xf_fname, "rb")
x0 <- readBin(con, "numeric", ns)
close(con)

xb <- matrix(rep(0, ns*nw), ncol=nw)
l2 <- rep(0, nc)

for (k in 1:nc){
  cat("cycle=", k, "\n")
  xt0 <- readBin(con.true, "numeric", ns)
  for (j in 2:nw){
      xt <- readBin(con.true, "numeric", ns)
  }
  yo <- matrix(readBin(con.obs, "numeric", ns*nw), nrow=ns)
  xb[, 1] <- x0
  for (i in 1:ni) {
    for (j in 2:nw){
      xb[, j] <- step.fom(l96, xb[, j-1], 1, dt, F)
    }
    d = xb - yo
    ad <- rep(0, ns)
    for (j in nw:1){
      ad <- step.adm(l96, l96.ad, xb[, j], ad, 1, dt, F) + d[, j] / r
    }
    xb[, 1] <- xb[, 1] - a * ad
    cost <- calc.cost(xb[, 1] - x0, b, d, r)
    if (i == 1) {
      cost.old <- cost
    }
    l2[k] <- sqrt(mean((xb[, 1] - xt0)^2))
    gnorm <- t(ad) %*% ad
    cat(sprintf("%d J=%e Jold=%e g=%e l2=%e\n", i, cost, cost.old, gnorm, l2[k]))
    if (gnorm < g.break) {
      break
    }
    if (cost > cost.old) {
      xb[, 1] <- xb[, 1] + a * ad
      l2[k] <- sqrt(mean((xb[, 1] - xt0)^2))
      break
    }
    cost.old <- cost
  }
  x0 <- step.fom(l96, xb[, 1], nw + 1, dt, F)
}
close(con.true)
close(con.obs)

con.out <- file("l96_var.dat", "wb")
writeBin(l2, con.out)
close(con.out)
